### PR TITLE
added dp toturial

### PR DESCRIPTION
**Pull request by ShizeSu06**
## Purpose
added DP tutorial documentation.

## What is the expected review turnaround time?

Urgency:
- [ ] High (needs review today)
- [x] Medium (needs review within a few days - most common)
- [ ] Low (can wait a week)

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# Differential privacy for cross-silo federated learning
+
+## Background
+Differential privacy (DP) is a popular and powerful technique that could add another layer of privacy protection for sensitive private data, with rigorous mathematical gurantee. DP technique can be applied to both Federated Learning (FL) applications as well as non-FL applications. There are different DP algorithms available in the literature, and among them DP-SGD algorithm is a well-known one that achieves State-of-The-Art performance in terms of model utility-privacy tradeoff. This tutorial focuses on the scenario about using DP-SGD for cross-silo FL.
```
Comment (garg-amit) [inline docs/tutorials/dp-for-cross-silo-fl.md:4]: nit: maybe `o` caps `State-of-The-Art`

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+### DP concept
+The theory for DP is complex, but the main idea is quite simple -- it clips datum gradient and adds Gaussian noise before model update to avoid privacy leakage. DP-SGD is an algorithm described in this [paper](https://arxiv.org/pdf/1607.00133.pdf); [Opacus](https://github.com/pytorch/opacus/blob/main/docs/faq.md) is its Pytorch implementation. Please refer to [this blog post](https://medium.com/pytorch/differential-privacy-series-part-1-dp-sgd-algorithm-explained-12512c3959a3) to read more about DP-SGD. 
+
+### How to apply DP in FL experiments
+When applying DP to cross-silo FL experiments, we just need to add DP-SGD to the silos training step, and the implementation using Opacus DP library is quite elegant. The example FL experiments in the current repo provide straigtforward and clear demonstration of using DP for cross-silo FL experiments, which you could easily mimic/adapt to your own cross-silo FL scenarios. Taking the fl_cross_silo_literal experiment using MNIST dataset as an example, briefly speaking, to implement DP-SGD, what you need are:
```
Comment (garg-amit) [inline docs/tutorials/dp-for-cross-silo-fl.md:13]: typo `straightforward`

**Review by garg-amit — APPROVED** [review]

File: docs/tutorials/dp-for-cross-silo-fl.md (hunk 1 above)
Comment (majercakdavid) [reply docs/tutorials/dp-for-cross-silo-fl.md:4]: nit: is this transferable to vertical fl? if not maybe we can specify that this targets horizontal only?

**Review by majercakdavid — APPROVED** [review]

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    # Total num of iterations
+    total_num_of_iterations=YAML_CONFIG.training_parameters.num_of_iterations,
+    ```
+It is also worthy to point out that, the provided DP cross-silo FL experiments in the current repo could be used for both DP-training and non-DP training. You can quickly turn DP off by setting the corresponding FL experiment `config.yaml` file's `dp` parameter to `false`, and turn DP on by setting `dp` parameter to `true`.
+
+### Tips and pitfalls about DP
+- DP privacy protection doesn't come without cost, and there is a model utility-privacy tradeoff. Epsilon is the DP privacy budget, and smaller epsilon means more privacy protection, more noises -- and hence worse model utility. For very sensitive scenarios, e.g., large language generative modeling tasks on very sensitive private data, DP might be desired to ensure approprite level of privacy protection. For less sensitive scenarios (e.g., standard classification modeling tasks), DP might be an overkill. Users should be aware of such model utility-privacy trade-off for DP, and evaluate their scenarios carefully to decide whether to apply DP, and if applying DP, with how much privacy budget.
```
Comment (jfomhover) [inline docs/tutorials/dp-for-cross-silo-fl.md:108]: This is a long section of text, can you add some structure here?

ex:

#### DP privacy doesn't come without cost

There is a model utility tradeoff...

#### Convergence is DP models is not trivial

...

etc

**Review by jfomhover — CHANGES_REQUESTED** [review]

File: docs/tutorials/dp-for-cross-silo-fl.md (hunk 1 above)
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md:4]: This dp tutorial focus only on cross-silo horizontal fl. I'd update accordingly.

File: docs/tutorials/dp-for-cross-silo-fl.md (hunk 2 above)
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md:13]: good catch!

File: docs/tutorials/dp-for-cross-silo-fl.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md:4]: good catch!

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+The theory for DP is complex, but the main idea is quite simple -- it clips datum gradient and adds Gaussian noise before model update to avoid privacy leakage. DP-SGD is an algorithm described in this [paper](https://arxiv.org/pdf/1607.00133.pdf); [Opacus](https://github.com/pytorch/opacus/blob/main/docs/faq.md) is its Pytorch implementation. Please refer to [this blog post](https://medium.com/pytorch/differential-privacy-series-part-1-dp-sgd-algorithm-explained-12512c3959a3) to read more about DP-SGD. 
+
+### How to apply DP in FL experiments
+When applying DP to cross-silo FL experiments, we just need to add DP-SGD to the silos training step, and the implementation using Opacus DP library is quite elegant. The example FL experiments in the current repo provide straigtforward and clear demonstration of using DP for cross-silo FL experiments, which you could easily mimic/adapt to your own cross-silo FL scenarios. Taking the fl_cross_silo_literal experiment using MNIST dataset as an example, briefly speaking, to implement DP-SGD, what you need are:
+- Adding DP libarary opacus dependencies to the [conda env yaml file](../../examples/components/MNIST/traininsilo/conda.yaml)
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md]: thanks for the suggestion! updated accordingly.

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        description: DP max gradient norm
+        default: 1.0
+        optional: true
+    total_num_of_iterations:
+        type: integer
+        description: Total num of iterations
+        default: 1
+        optional: true
+    ```
+- Define privacy engine and wrap model training with DP in the corresponding silo [training component run.py](../../examples/components/MNIST/traininsilo/run.py) file, like below. Compared with standard non-FL DP training, there is one thing noteworthy about `privacy_engine.make_private_with_epsilon()`, cross-silo FL experiments would include multiple iterations of silo training, and we need to specify `epochs=total_num_of_iterations * epochs` to correctly apply the right level of noise for a given privacy budget (dp_target_epsilon) for the final DP-trained FL model. Some model architecture layers might be incompatible with Opacus DP library, and Opacus ModuleValidator is helpful to validate the model architecture compatiblity and help fix/replace the incompatible layers, shown as below.
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md]: thanks for the suggestion! updated accordingly.

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                optimizer=self.optimizer_,
+                data_loader=self.train_loader_,
+                epochs=total_num_of_iterations * epochs,
+                target_epsilon=dp_target_epsilon,
+                target_delta=dp_target_delta,
+                max_grad_norm=dp_max_grad_norm,
+            )
+
+            """
+            Alternatively, you can also obtain their counterparts by passing the noise multiplier, if you don't have a targeted privacy budget epsilon in mind, and prefer to play with different level of DP noise multiplier.
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md]: updated.

File: docs/tutorials/dp-for-cross-silo-fl.md (hunk 3 above)
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md]: updated.

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            privacy_engine.make_private(
+                module=self.model_,
+                optimizer=self.optimizer_,
+                data_loader=self.train_loader_,
+                noise_multiplier=dp_noise_multiplier,
+                max_grad_norm=dp_max_grad_norm,
+            )
+            """
+    ```
+    ```python
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md]: updated.

File: docs/tutorials/dp-for-cross-silo-fl.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        type: integer
+        description: Total num of iterations
+        default: 1
+        optional: true
+    ```
+- Define privacy engine and wrap model training with DP in the corresponding silo [training component run.py](../../examples/components/MNIST/traininsilo/run.py) file, like below. Compared with standard non-FL DP training, there is one thing noteworthy about `privacy_engine.make_private_with_epsilon()`, cross-silo FL experiments would include multiple iterations of silo training, and we need to specify `epochs=total_num_of_iterations * epochs` to correctly apply the right level of noise for a given privacy budget (dp_target_epsilon) for the final DP-trained FL model. Some model architecture layers might be incompatible with Opacus DP library, and Opacus ModuleValidator is helpful to validate the model architecture compatiblity and help fix/replace the incompatible layers, shown as below.
+    ```python
+    if dp:
+            privacy_engine = PrivacyEngine(secure_mode=False)
+            """secure_mode: Set to True if cryptographically strong DP guarantee is
```
[the remark this answers is not in the document]
Comment (ShizeSu06) [reply docs/tutorials/dp-for-cross-silo-fl.md]: updated.

**Review by jfomhover — APPROVED** [review]
